lib/query/event.js

```diff
@@ -18,52 +20,80 @@ function buildEventQuery (params) {
   const {
     metric,
     dimension,
-    startDate,
-    endDate,
+    time: {
+      startDate,
+      endDate
+    },
     where
   } = params
 
   const rawWhere = buildRawWhere({ startDate, endDate, where })
 
   if (isDimensionTime(dimension)) {
-    return buildTimeDimensionedEventCountQuery({ metric, dimension, rawWhere })
+    return buildTimeDimensionedEventCountQuery({ metric, dimension, rawWhere, startDate, endDate })
   }
 
   if (dimension) {
     return buildDimensionedEventCountQuery({ metric, dimension, rawWhere })
   }
 
+  const query = buildEventCountQuery({ metric, rawWhere })
+  console.log(query.toString())
   return buildEventCountQuery({ metric, rawWhere })
 }
 
-function buildTimeDimensionedEventCountQuery ({ metric, dimension, rawWhere }) {
-  return knex.select(knex.raw(`DATE_TRUNC('${dimension}', event_timestamp ) AS timestamp`))
-    .count(`${eventColumn} AS ${metric}`)
+function buildTimeDimensionedEventCountQuery ({ metric, dimension, rawWhere, startDate, endDate }) {
+  const metricAlias = getMetricAlias(metric)
+  return db.select(db.raw(`DATE_TRUNC('${dimension}', ${timestampColumn} ) AS timestamp`))
+    .count(`${eventColumn} AS ${metricAlias}`)
     .withSchema(schema)
     .from(table)
     .where(eventColumn, '=', eventLookup[metric])
     .andWhereRaw(rawWhere)
-    .groupByRaw(`DATE_TRUNC('${dimension}', event_timestamp )`)
-    .orderByRaw(`DATE_TRUNC('${dimension}', event_timestamp )`)
+    .groupByRaw(`DATE_TRUNC('${dimension}', ${timestampColumn} )`)
+    .orderByRaw(`DATE_TRUNC('${dimension}', ${timestampColumn} )`)
```

Comment:
if dimension comes in from the user, should make sure it's properly escaped.

Reply:
The dimension does come from the user, but we use Joi to validate that its value is one from a list of valid values. Any request with a dimension not in the list would get rejected.
This pattern is a bit inflexible - anytime you want to support a new dimension, it takes a change to configuration settings.  It does however prevent unwanted dimensioning (i.e., expensive queries) against the Redshift instance.

Reply:
cool. just wanted to guard against sql injection.